bundles/async_remote_services/admin/src/admin.cc

```diff
@@ -21,300 +21,256 @@
-            cppProps.set(key, celix_properties_get(props, key, nullptr));
-        }
-
-        _toBeCreatedExportedEndpoints.emplace_back(std::make_pair(svc, std::move(cppProps)));
-        return;
-    }
-
-    _exportedServiceInstances.emplace(endpointId, factory->second->create(svc, endpointId));
+    std::lock_guard<std::mutex> lock{_m};
+    _toBeExportedServices.emplace_back(svc, props);
+    createExportServices();
 }
 
-void celix::async_rsa::AsyncAdmin::removeService(void *, const celix_properties_t *props) {
-    auto *objectClass = celix_properties_get(props, OSGI_FRAMEWORK_OBJECTCLASS, nullptr);
-    auto *remote = celix_properties_get(props, "remote", nullptr);
-    auto svcId = celix_properties_get(props, ENDPOINT_IDENTIFIER, nullptr);
-
-    if(objectClass == nullptr) {
-        L_WARN("Removing service to be exported but missing objectclass");
-        return;
-    }
+void celix::async_rsa::AsyncAdmin::removeService(const std::shared_ptr<void>& /*svc*/, const std::shared_ptr<const celix::Properties>& props) {
+    assert(props->getAsBool(celix::rsa::REMOTE_SERVICE_PROPERTY_NAME, false)); //"only remote services should be tracker by the service tracker");
 
-    if(remote == nullptr) {
-        // not every service is remote, this is fine but not something the RSA admin is interested in.
+    long svcId = props->getAsLong(celix::SERVICE_ID, -1);
+    if (svcId < 0) {
+        L_WARN("Removing service to be exported but missing service.id");
         return;
     }
 
-    if(svcId == nullptr) {
-        L_WARN("Removing service to be exported but missing endpoint.id");
-        return;
-    }
+    std::lock_guard l(_m);
 
-    std::unique_lock l(_m);
+    //remove export interface (if present)
     auto instanceIt = _exportedServiceInstances.find(svcId);
-
-    if(instanceIt == end(_exportedServiceInstances)) {
-        return;
+    if (instanceIt != end(_exportedServiceInstances)) {
+        //note cannot remove async, because the svc lifetime needs to extend the lifetime of the component.
+        //TODO improve by creating export service based on svc id instead of svc pointer, so that export service can track its own dependencies.
+        _ctx->getDependencyManager()->removeComponent(instanceIt->second);
+        _exportedServiceInstances.erase(instanceIt);
     }
 
-    _mng->destroyComponent(instanceIt->second);
-    _exportedServiceInstances.erase(instanceIt);
-}
-
-void celix::async_rsa::AsyncAdmin::addEndpointInternal(celix::rsa::Endpoint& endpoint) {
-
-    const auto& properties = endpoint.getProperties();
-    auto interface = properties.get(ENDPOINT_EXPORTS);
-
-    if(interface.empty()) {
-        L_WARN("Adding endpoint but missing exported interfaces");
-        return;
+    //remove to be exported endpoint (if present)
+    for (auto it = _toBeExportedServices.begin(); it != _toBeExportedServices.end(); ++it) {
+        if (it->second->getAsBool(celix::SERVICE_ID, -1) == svcId) {
```

Comment:
`getAsLong` is probably a better match here instead of `getAsBool`

Reply:
nice catch